src/app/qgscrashreport.cpp

```diff
@@ -0,0 +1,149 @@
+
+QgsCrashReport::QgsCrashReport()
+{
+  setFlags( QgsCrashReport::Stack |
+            QgsCrashReport::Plugins |
+            QgsCrashReport::ProjectDetails |
+            QgsCrashReport::SystemInfo |
+            QgsCrashReport::QGISInfo );
+}
+
+void QgsCrashReport::setFlags( QgsCrashReport::Flags flags )
+{
+  mFlags = flags;
+}
+
+QString QgsCrashReport::toString()
+{
+  QStringList reportData;
+  reportData.append( "Crash ID: " + crashID() );
+
+  if ( flags().testFlag( QgsCrashReport::Stack ) )
+  {
+    reportData.append( "\n" );
+    reportData.append( "Stack Trace" );
+    reportData.append( "===========" );
+    Q_FOREACH ( QgsCrashReport::StackLine line, mStackTrace )
+    {
+      QFileInfo fileInfo( line.FileName );
+      QString filename( fileInfo.fileName() );
+      reportData.append( QString( "(%1) %2 %3:%4" ).arg( line.ModuleName, line.SymbolName, filename, line.LineNumber ) );
+    }
+  }
+
+  if ( flags().testFlag( QgsCrashReport::Plugins ) )
+  {
+    reportData.append( "\n" );
+    reportData.append( "Plugins" );
+    reportData.append( "=======" );
+    // TODO Get plugin info
+  }
+
+  if ( flags().testFlag( QgsCrashReport::ProjectDetails ) )
+  {
+    reportData.append( "\n" );
+    reportData.append( "Project Info" );
+    reportData.append( "============" );
+    // TODO Get project details
+  }
+
+  if ( flags().testFlag( QgsCrashReport::QGISInfo ) )
+  {
+    reportData.append( "\n" );
+    reportData.append( "QGIS Info" );
+    reportData.append( "=========" );
+    reportData.append( QString( "QGIS Version: %1" ).arg( Qgis::QGIS_VERSION ) );
+
+    if ( QString( Qgis::QGIS_DEV_VERSION ) == QLatin1String( "exported" ) )
+    {
+      reportData.append( QString( "QGIS code branch: Release %1.%2" )
+                         .arg( Qgis::QGIS_VERSION_INT / 10000 ).arg( Qgis::QGIS_VERSION_INT / 100 % 100 ) );
+    }
+    else
+    {
+      reportData.append( QString( "QGIS code revision: %1" ).arg( Qgis::QGIS_DEV_VERSION ) );
+    }
+
+    reportData.append( QString( "Compiled agaist Qt: %1" ).arg( QT_VERSION_STR ) );
+    reportData.append( QString( "Running agaist Qt: %1" ).arg( qVersion() ) );
+
+    reportData.append( QString( "Compiled agaist GDAL: %1" ).arg( GDAL_RELEASE_NAME ) );
+    reportData.append( QString( "Running agaist GDAL: %1" ).arg( GDALVersionInfo( "RELEASE_NAME" ) ) );
+  }
+
+  if ( flags().testFlag( QgsCrashReport::SystemInfo ) )
+  {
+    reportData.append( "\n" );
+    reportData.append( "System Info" );
+    reportData.append( "===========" );
+    reportData.append( QString( "CPU Type: %1" ).arg( QSysInfo::currentCpuArchitecture() ) );
+    reportData.append( QString( "Kernal Type: %1" ).arg( QSysInfo::kernelType() ) );
+    reportData.append( QString( "Kernal Version: %1" ).arg( QSysInfo::kernelVersion() ) );
+  }
+
+  QString report;
+  Q_FOREACH ( QString line, reportData )
+  {
+    report += line + "\n";
+  }
+  return report;
+}
+
+QString QgsCrashReport::crashID()
```

Comment:
I think I would prefer ID out of the full stack trace... often top of the trace may be something inside Qt code and we may wrongly identify different crashes as the same one. I would not be worried that things would be too easy to change and thus produce different crash IDs for the same problem - I guess most of the time users will use one of the few released versions where the crash IDs should be stable...